man/amrmtape.8
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "AMRMTAPE" 8 "" "" ""
.SH NAME
amrmtape \- remove a tape from the AMANDA database
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 9
\fBamrmtape\fR [\-n] [\-v] [\-q] [\-d] \fIconfig\fR \fIlabel\fR
.ad
.hy

.SH "DESCRIPTION"

.PP
\fBAmrmtape\fR invalidates the contents of an existing backup tape in the configuration database\&. This is meant as a recovery mechanism when a good backup is damaged either by faulty hardware or user error, e\&.g\&. the tape is eaten by the drive or is overwritten\&.

.PP
See the \fBamanda\fR(8) man page for more details about \fBAMANDA\fR\&.

.SH "OPTIONS"

.TP
\fB\-n\fR
Generate new \fBtapelist\fR and database files with \fBlabel\fR removed, but leave them in \fI/tmp\fR and do not update the original copies\&.

.TP
\fB\-v\fR
List backups of hosts and disks that are being discarded\&. Enabled by default\&.

.TP
\fB\-q\fR
Opposite of \fB\-v\fR\&.

.TP
\fB\-d\fR
Run in debugging mode so all executed commands are shown\&.

.SH "EXAMPLE"

.PP
Remove tape labeled \fBDAILY034\fR from the \fBDailySet1\fR configuration\&.
.nf

# amrmtape DailySet1 DAILY034
.fi

.SH "AUTHOR"

.PP
Adrian T\&. Filipi\-Martin <atf3r@cs\&.virginia\&.edu>: Original text

.PP
Stefan G\&. Weichinger, <sgw@amanda\&.org>, maintainer of the \fBAMANDA\fR\-documentation: XML\-conversion

.SH "SEE ALSO"

.PP
\fBamadmin\fR(8), \fBamanda\fR(8)



NEWS
Changes in release 2.5.1b2

* debug file are seperated in subdir client/server/amandad with the config name.
* amandates client config option.
* gnutar_list-dir client config option.
* works on solaris 10 64bits.

Changes in release 2.5.1b1

* Klocwork defects fixed.
* Coverity defect fixed.
* work with tar-1.15.91
* Open SSL encryption support
* amfetchdump -o is replaced by -O.
* amrecover_do_fsf default to yes
* amrecover_check_label default to yes
* amcheck -w will do all test including the tape writable test.
  use amcheck -t -w if you want only the tape writable test.
* usetimestamps
* bsdudp/bsdtcp, 2 new auth.
* Unlimited number of DLE on a client with bsdtcp/rsh or ssh auth.
* docs/howto-auth.txt
* amandad_path, client_username and ssh_keys (for ssh/rsh)
* client config file (see man amanda-client.conf)
  Only use by amrecover.
* amrecover use the security-api
* amoldrecover which use the old amindexd/amidxtaped protocol
* -o option to overwrite config option
  (see in man amanda, the CONFIGURATION OVERWRITE section)
* new format of .amandahosts
* new format of xinetd.d/amanda

Changes in release 2.5.0p2

* listhost subcommand in amrecover

Changes in release 2.5.0p1

* Add the 'amtape update' command.

Changes in release 2.5.0

* Communication security/authentication: Kerberos 4/5, OpenSSH
* Data security: Symmetric/Assymetric encrytion algorithms (aesutil
  and gpg encryption),  Encryption can be done on server or client,
  Custom encryption utilities can be used.
* Compression: Ability to add custom compression tools. This is a
  really useful feature especially since it can specified for a DLE.  You
  can use different compression algorithm for images, binaries, ascii files
  and so on.
* Dump images spanning multiple media volumes - Dump images are no
  longer restricted to a single media volume (tape or vtape).  Data restoration
  can be done using amrecover and amfetchdump commands.
* Auto tape labelling - This optional feature is good for disk backups.

Change in release 2.4.5p1

* New [--days <num>] argument to 'amadmin <conf> tape'.
* New amoverview -last -num0 -togo0 options.

Change in release 2.4.5

* new displayunit global option to select the unit use to display number
  k=kilo, m=mega, g=giga, t=tera.
* new amoverview -skipmissed option.

Change in release 2.4.5b1

* holding disk disk use timestamped directory.
* autoflush flush today's dump.
* new bumppercent global option, this should improve bumping criteria,
  the bumpsize is set to a fixed value which can be adequate for small
  and large disk.
* bumpsize, bumppercent, bumpdays and bumpmult can be in a dumptype.
* calcsize support include and exclude like gnutar.
* new 'estimate' dumptype option to select estimate type:
    CLIENT: estimate by the dumping program.
    CALCSIZE: estimate by the calcsize program, a lot faster but less acurate.
    SERVER: estimate based on statistic from previous run, take second but
            can be wrong on the estimate size.

Changes in release 2.4.4p3

* new chg-iomega changer script.
* amanda will not use a tape if it's label is not in the tapelist file
  (eg. after an amrmtape).

Changes in release 2.4.4p2

* new initial_poll_delay config in chg-zd-mtx
* 'amadmin find' list disk removed from the disklist.
* amrecover can recover a disk removed from the disklist file.
* amanda works with samba3.
* new chg-disk changer script to use with the file: driver.
* Kerberos 4 support integrated and functional again.

Changes in release 2.4.4p1

* amverify do not advance to next tape at the end.
* new amflush -b and -s options, change -f behavior.
* new chg-mcutil changer.
* amrecover_changer works with chg-multi.
* default driveslot for chg-zd-mtx is now 0.
* amplot generate color postscript by default.
* amplot -b flag to generate b/w postscript.
* rait: works with file:
* print a postscript label for each tape used.

Changes in release 2.4.4

* New -a option to amcheck to always send an email.

Changes in release 2.4.4b1

* maxpromoteday: New option for a dumptype.
* New amtapetype program (replace tapetype)
* Client compile on CYGWIN
* amrestore -f <fileno> : amrestore can do fsf before reading the tape
* amrestore -l <label>  : amrestore can check the label before restoring.
* New config options:
  amrecover_do_fsf: Amrecover will use the -f flag of amrestore to position the
                    tape.
  amrecover_check_label: amrecover will pass the -l flag to amrestore.
  amrecover_changer: amrecover will use your changer.
  maxdumpsize: The maximum size of the dumps during a run,
               default to tapesize * runtapes
  taperalgo: select your algo for the taper, try 'largestfit', it can improve
             tape usage.
* amrecover allow to retry a skip a tape.
* New --days option to 'amadmin <config> balance' command.
* New --date option to amstatus
* amreport print an usage by tape.
* Newer promote algorithm that try to reduce the number of full on a
  single host in a run.
* New changer: chg-juke, chg-null and chg-rait

Changes in release 2.4.3

* Few small bug fixed
* New runtapes argument to amverify.
* New amverifyrun program.
* 2.4.3 client works with older server.

Changes in release 2.4.3b4

* new noop packet type. It allow the server to know the client features.
* exclude list with relative path is working again.
* amstatus is working again.

Changes in release 2.4.3b3

* --with-maxtapeblocksize configure options
* blocksize tapetype option
* file-pad tapetype option
* Multiple exclude in dumptype
* Option include in dumptype
* New disklist syntax:
*   hostname diskname [ diskdevice ] dumptype [ spindle [ interface ] ]
* chg-zd-mtx: Major cleanup and general overhaul.
* amrecover: new listdisk command.

Changes in release 2.4.3b2

* amrestore try to find the next chunk in the current directory.
* amrecover: You can change the tape device between each tape.
* amstatus work with amflush.

Changes in release 2.4.3b1

* tapeio.

server-src/amrmtape.pl
#!@PERL@
#
# Copyright (c) 2008, 2009, 2010 Zmanda, Inc.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
#
# Contact information: Zmanda Inc, 465 S. Mathilda Ave., Suite 300
# Sunnyvale, CA 94086, USA, or: http://www.zmanda.com
#
use lib '@amperldir@';
use strict;
use warnings;

use Amanda::Device qw( :constants );
use Amanda::Config qw( :init :getconf config_print_errors
  config_dir_relative new_config_overrides add_config_override);
use Amanda::Changer;
use Amanda::Debug qw( :logging );
use Amanda::Disklist;
use Amanda::Paths;
use Amanda::MainLoop;
use Amanda::Tapelist;
use Amanda::Util qw( :constants );
use File::Copy;
use File::Basename;
use Getopt::Long;

my $amadmin = "$sbindir/amadmin";
my $amtrmidx = "$amlibexecdir/amtrmidx";
my $amtrmlog = "$amlibexecdir/amtrmlog";

my $dry_run;
my $cleanup;
my $erase;
my $changer_name;
my $keep_label;
my $verbose = 1;
my $help;
my $logdir;
my $log_file;
my $log_created = 0;

sub die_handler {
    if ($log_created == 1) {
        unlink $log_file;
        $log_created = 0;
    }
}
$SIG{__DIE__} = \&die_handler;

sub int_handler {
    if ($log_created == 1) {
        unlink $log_file;
        $log_created = 0;
    }
    die "Interrupted\n";
}
$SIG{INT} = \&int_handler;

sub usage() {
    print <<EOF
$0 [-n] [-v] [-q] [-d] [config-overwrites] <config> <label>
\t--changer changer-name
\t\tSpecify the name of the changer to use (for --erase).
\t--cleanup
\t\tRemove indexes and logs immediately
\t-n, --dryrun
\t\tDo nothing to original files, leave new ones in database directory.
\t--erase
\t\tErase the media, if possible
\t-h, --help
\t\tDisplay this message.
\t--keep-label
\t\tDo not remove label from the tapelist
\t-q, --quiet
\t\tQuiet, opposite of -v.
\t-v, --verbose
\t\tVerbose, list backups of hosts and disks that are being discarded.

This program allows you to invalidate the contents of an existing
backup tape within the Amanda current tape database.  This is meant as
a recovery mecanism for when a good backup is damaged either by faulty
hardware or user error, i.e. the tape is eaten by the tape drive, or
the tape has been overwritten.
EOF
}

sub vlog(@) {
    foreach my $msg (@_) {
        message($msg);
        print "$0: $msg\n" if $verbose;
    }
}

Amanda::Util::setup_application("amrmtape", "server", $CONTEXT_CMDLINE);

my $config_overrides = new_config_overrides( scalar(@ARGV) + 1 );

my $opts_ok = GetOptions(
    "changer=s" => \$changer_name,
    "cleanup" => \$cleanup,
    "dryrun|n" => \$dry_run,
    "erase" => \$erase,
    "help|h" => \$help,
    "keep-label" => \$keep_label,
    'o=s' => sub { add_config_override_opt( $config_overrides, $_[1] ); },
    "quiet|q" => sub { undef $verbose; },
    "verbose|v" => \$verbose,
);

unless ($opts_ok && scalar(@ARGV) == 2) {
    unless (scalar(@ARGV) == 2) {
        print STDERR "Specify a configuration and label.\n";
    }
    usage();
    exit 1;
}

if ($help) {
    usage();
    exit 0;
}

my ($config_name, $label) = @ARGV;

set_config_overrides($config_overrides);
my $cfg_ok = config_init( $CONFIG_INIT_EXPLICIT_NAME, $config_name );

my ($cfgerr_level, @cfgerr_errors) = config_errors();
if ($cfgerr_level >= $CFGERR_WARNINGS) {
    config_print_errors();
    if ($cfgerr_level >= $CFGERR_ERRORS) {
        die "Errors processing config file";
    }
}

Amanda::Util::finish_setup($RUNNING_AS_DUMPUSER);
$logdir = config_dir_relative(getconf($CNF_LOGDIR));
$log_file = "$logdir/log";

if ($erase) {
    # Check for log file existance
    if (-e $log_file) {
        `amcleanup -p $config_name`;
    }

    if (-e $log_file) {
        local *LOG;
        open(LOG,  $log_file);
        my $info_line = <LOG>;
        close LOG;
        $info_line =~ /^INFO (.*) .* pid .*$/;
        my $process_name = $1;
        print "$process_name is running, or you must run amcleanup\n";
        exit 1;
    }
}

# amadmin may later try to load this and will die if it has errors
# load it now to catch the problem sooner (before we might erase data)
my $diskfile = config_dir_relative(getconf($CNF_DISKFILE));
$cfgerr_level = Amanda::Disklist::read_disklist('filename' => $diskfile);
if ($cfgerr_level >= $CFGERR_ERRORS) {
    die "Errors processing disklist";
}

my $scrub_db = sub {
    my $tapelist_file = config_dir_relative(getconf($CNF_TAPELIST));
    my $tapelist = Amanda::Tapelist::read_tapelist($tapelist_file);
    unless ($tapelist) {
        die "Could not read the tapelist";
    }

    my $t = $tapelist->lookup_tapelabel($label);
    if ($keep_label) {
        $t->{'datestamp'} = 0 if $t;
    } elsif (!defined $t) {
	print "label '$label' not found in $tapelist_file\n";
	exit 0;
    } else {
        $tapelist->remove_tapelabel($label);
    }

    #take a copy in case we roolback
    my $backup_tapelist_file = dirname($tapelist_file) . "-backup-amrmtape-" . time();
    if (-x $tapelist_file) {
	unless (copy($tapelist_file, $backup_tapelist_file)) {
	    die "Failed to copy/backup $tapelist_file to $backup_tapelist_file";
	}
    }

    unless ($dry_run) {
        $tapelist->write($tapelist_file);
    }

    my $tmp_curinfo_file = "$AMANDA_TMPDIR/curinfo-amrmtape-" . time();
    unless (open(AMADMIN, "$amadmin $config_name export |")) {
        die "Failed to execute $amadmin: $! $?";
    }
    open(CURINFO, ">$tmp_curinfo_file");

    sub info_line($) {
        print CURINFO "$_[0]";
    }

    my $host;
    my $disk;
    my $dead_level = 10;
    while(my $line = <AMADMIN>) {
        my @parts = split(/\s+/, $line);
        if ($parts[0] =~ /^CURINFO|#|(?:command|last_level|consecutive_runs|(?:full|incr)-(?:rate|comp)):$/) {
            info_line $line;
        } elsif ($parts[0] eq 'host:') {
            $host = $parts[1];
            info_line $line;
        } elsif ($parts[0] eq 'disk:') {
            $disk = $parts[1];
            info_line $line;
        } elsif ($parts[0] eq 'history:') {
            info_line $line;
        } elsif ($line eq "//\n") {
            info_line $line;
            $dead_level = 10;
        } elsif ($parts[0] eq 'stats:') {
            if (scalar(@parts) < 6 || scalar(@parts) > 8) {
                die "unexpected number of fields in \"stats\" entry for $host:$disk\n\t$line";
            }
            my $level = $parts[1];
            my $cur_label = $parts[7];
            if ($cur_label eq $label) {
                $dead_level = $level;
                vlog "Discarding Host: $host, Disk: $disk, Level: $level\n";
            } elsif ( $level > $dead_level ) {
                vlog "Discarding Host: $host, Disk: $disk, Level: $level\n";
            } else {
                info_line $line;
            }
        } else {
            die "Error: unrecognized line of input:\n\t$line";
        }
    }

    my $rollback_from_curinfo = sub {
            unlink $tmp_curinfo_file;
            return if $keep_label;
            unless (move($backup_tapelist_file, $tapelist_file)) {
                printf STDERR "Failed to rollback new tapelist.\n";
            }
    };

    close CURINFO;

    unless (close AMADMIN) {
        $rollback_from_curinfo->();
        die "$amadmin exited with non-zero while exporting: $! $?";
    }

    unless ($dry_run) {
        if (system("$amadmin $config_name import < $tmp_curinfo_file")) {
            $rollback_from_curinfo->();
            die "$amadmin exited with non-zero while importing: $! $?";
        }
    }

    unlink $tmp_curinfo_file;
    unlink $backup_tapelist_file;

    if ($cleanup && !$dry_run) {
        if (system($amtrmlog, $config_name)) {
            die "$amtrmlog exited with non-zero while scrubbing logs: $! $?";
        }
        if (system($amtrmidx, $config_name)) {
            die "$amtrmidx exited with non-zero while scrubbing indexes: $! $?";
        }
    }

    Amanda::MainLoop::quit();
};

my $erase_volume = make_cb('erase_volume' => sub {
    if ($erase) {
        $log_created = 1;
        local *LOG;
        open(LOG, ">$log_file");
        print LOG "INFO amrmtape amrmtape pid $$\n";
        close LOG;
	my $chg = Amanda::Changer->new($changer_name);
	$chg->load(
	    'label' => $label,
	    'res_cb' => sub {
		my ($err, $resv) = @_;
		die $err if $err;

                my $dev = $resv->{'device'};
                die "Can not erase $label because the device doesn't support this feature"
                    unless $dev->property_get('full_deletion');
                if (!$dry_run) {
                    $dev->erase()
                        or die "Failed to erase volume";
                    $dev->finish();

                    # label the tape with the same label it had
                    if ($keep_label) {
                        $dev->start($ACCESS_WRITE, $label, undef)
                            or die "Failed to write tape label";
                    }
                }

		$resv->release(finished_cb => sub {
		    my ($err) = @_;
		    die $err if $err;

		    $scrub_db->();
		});
            });
    } else {
        $scrub_db->();
    }
});

# kick things off
$erase_volume->();
Amanda::MainLoop::run();

if ($log_created == 1) {
    unlink $log_file;
    $log_created = 0;
}

Amanda::Util::finish_application();
